frontend/src/metabase-lib/lib/queries/utils/virtual-table.ts

```diff
@@ -0,0 +1,65 @@
+import Table from "metabase-lib/lib/metadata/Table";
+import Field from "metabase-lib/lib/metadata/Field";
+import type Database from "metabase-lib/lib/metadata/Database";
+import type Metadata from "metabase-lib/lib/metadata/Metadata";
+import type StructuredQuery from "../StructuredQuery";
+import type NativeQuery from "../NativeQuery";
+
+type VirtualTableProps = {
+  metadata: Metadata;
+  fields?: Field[];
+  db?: Database | null;
+} & Partial<Table>;
+
+type VirtualFieldProps = {
+  metadata: Metadata;
+  query: StructuredQuery | NativeQuery;
+} & Partial<Field>;
+
+export function createVirtualTable({
+  metadata,
+  fields,
+  db,
+  ...rest
+}: VirtualTableProps): Table {
+  const table = new Table({
+    name: "",
+    display_name: "",
+    ...rest,
+  });
+
+  Object.assign(table, {
+    fields: fields || [],
+    db,
+    metadata,
+    segments: [],
+    metrics: [],
```

Comment:
There's nothing stopping all the props that go in the object arg of the constructor from going onto the underlying `plainObject`, so trying my best here to avoid that.